File: LG_ArchJ/app/src/main/java/com/louis/lg_archj/ui/news/NewsFragment.java
package com.louis.lg_archj.ui.news;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import com.louis.lg_archj.di.AppContainer;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.louis.lg_archj.databinding.FragmentNewsBinding;

/**
 * A simple {@link Fragment} subclass.
 * Use the {@link NewsFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class NewsFragment extends Fragment {
    private static final String ARG_PARAM1 = "param1";
    private static final String ARG_PARAM2 = "param2";

    private String mParam1;
    private String mParam2;
    private FragmentNewsBinding binding;
    private NewsViewModel viewModel;
    private NewsAdapter newsAdapter;

    public NewsFragment() {
    }

    public static NewsFragment newInstance(String param1, String param2) {
        NewsFragment fragment = new NewsFragment();
        Bundle args = new Bundle();
        args.putString(ARG_PARAM1, param1);
        args.putString(ARG_PARAM2, param2);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            mParam1 = getArguments().getString(ARG_PARAM1);
            mParam2 = getArguments().getString(ARG_PARAM2);
        }

        // 初始化 ViewModel 使用依赖注入
        AppContainer container = AppContainer.getInstance();
        NewsViewModelFactory factory = new NewsViewModelFactory(container.getLoadNewsUseCase());
        viewModel = new ViewModelProvider(this, factory).get(NewsViewModel.class);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = FragmentNewsBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        // 初始化 RecyclerView
        initRecyclerView();

        // 设置下拉刷新监听器
        binding.swipeRefreshLayout.setOnRefreshListener(() -> {
            viewModel.processIntent(new NewsUiIntent.RefreshData());
        });

        // 观察 UI 状态变化
        viewModel.uiState.observe(getViewLifecycleOwner(), this::render);

        // 发送初始加载数据的意图
        viewModel.processIntent(new NewsUiIntent.LoadData());
    }

    private void initRecyclerView() {
        newsAdapter = new NewsAdapter();
        binding.newsRecyclerView.setLayoutManager(new LinearLayoutManager(requireContext()));
        binding.newsRecyclerView.setAdapter(newsAdapter);
    }

    private void render(NewsUiState state) {
        if (state.isLoading()) {
            //显示加载状态
            binding.progressBar.setVisibility(View.VISIBLE);
            binding.swipeRefreshLayout.setRefreshing(false);
            binding.newsRecyclerView.setVisibility(View.GONE);
            binding.errorText.setVisibility(View.GONE);
        } else if (state.isRefreshing()) {
            //显示刷新状态
            binding.progressBar.setVisibility(View.GONE);
            binding.swipeRefreshLayout.setRefreshing(true);
            binding.newsRecyclerView.setVisibility(View.VISIBLE);
            binding.errorText.setVisibility(View.GONE);
        } else if (state.error != null) {
            // 显示错误状态
            binding.progressBar.setVisibility(View.GONE);
            binding.swipeRefreshLayout.setRefreshing(false);
            binding.newsRecyclerView.setVisibility(View.GONE);
            binding.errorText.setVisibility(View.VISIBLE);
            binding.errorText.setText(state.error);
        } else if (state.data != null) {
            // 显示数据
            binding.progressBar.setVisibility(View.GONE);
            binding.swipeRefreshLayout.setRefreshing(false);
            binding.newsRecyclerView.setVisibility(View.VISIBLE);
            binding.errorText.setVisibility(View.GONE);

            newsAdapter.updateNews(state.data);
        }
    }
}

File: MyArchitecture/app/src/main/java/com/louis/myarchitecture/ui/NewsFragment.kt
package com.louis.myarchitecture.ui

import androidx.fragment.app.viewModels
import android.os.Bundle
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.Observer
import androidx.lifecycle.flowWithLifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.louis.myarchitecture.R
import com.louisgeek.library.tool.ToastTool
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch

class NewsFragment : Fragment() {

    companion object {
        fun newInstance() = NewsFragment()
    }

    private val viewModel: NewsViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        return inflater.inflate(R.layout.fragment_news, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val tv: TextView = view.findViewById(R.id.tv)
        tv.setOnClickListener {
            viewModel.clickTv()
        }

        viewModel.uiEvent.observe(viewLifecycleOwner) { uiEvent ->
            when (uiEvent) {
                is NewsUIEvent.ShowMessage -> {
                    Toast.makeText(requireContext(), uiEvent.message, Toast.LENGTH_SHORT).show()
                }

            }

        }

        viewModel.uiState.observe(viewLifecycleOwner) { uiState ->
            when (uiState) {
                is NewsUiState.Loading -> {

                }

                is NewsUiState.Success -> {
                    uiState.news
                }

                is NewsUiState.Error -> {
                    uiState.msg
                }
            }

        }

        lifecycleScope.launch {
            lifecycle.repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.uiStateFlow //StateFlow
                    .map { it } // Flow
                    .distinctUntilChanged() //过滤重复元素
                    .collect { uiState ->
                        when (uiState) {
                            is NewsUiState.Loading -> {}
                            is NewsUiState.Success -> {

                            }

                            is NewsUiState.Error -> {

                            }
                        }
                    }
            }
        }

        lifecycleScope.launch {
            viewModel.uiStateFlow
                .flowWithLifecycle(lifecycle, Lifecycle.State.STARTED) //简化代码
                .collect { uiState ->
                    when (uiState) {
                        is NewsUiState.Loading -> {}
                        is NewsUiState.Success -> {

                        }

                        is NewsUiState.Error -> {

                        }
                    }
                }
        }

        viewModel.fetchNews()

    }
}